src/foam/core/session/PersistentSessionDAO.js

```diff
@@ -0,0 +1,165 @@
+ */
+
+foam.CLASS({
+  package: 'foam.core.session',
+  name: 'PersistentSessionDAO',
+  extends: 'foam.dao.ProxyDAO',
+  implements: [
+    'foam.core.COREService'
+  ],
+
+  documentation: `The key to 'persistent' sessions is the userId on the
+session. If a user makes a client call with an existing session id and
+a matching user id, then the system considers the user logged in.
+Sessions are NOT-frozen, for reasons discussed elsewhere.  The
+implication is updates are not persisted to the runtime journal because
+there is never a delta.
+This DAO maintains it's own copy of the session to calculate the delta
+and then makes an explicit journal put call to store the session with
+the user id.
+
+Additionally, anonymous also adds a complication.  When a user is logged
+out, the client will present anonymous views and the anonymous user
+uses the same session id as it is the same client connection. This DAO
+also ensures the session is not saved with the anonymous user id.
+`,
+
+  javaImports: [
+    'foam.core.auth.AuthService',
+    'foam.core.auth.ServiceProvider',
+    'foam.core.logger.Logger',
+    'foam.core.logger.Loggers',
+    'foam.dao.ArraySink',
+    'foam.dao.DAO',
+    'foam.dao.java.JDAO',
+    'foam.dao.Journal',
+    'foam.dao.MDAO',
+    'foam.dao.NullDAO',
+    'foam.dao.ProxyDAO',
+    'foam.lang.ContextAgent',
+    'foam.lang.ContextAgentTimerTask',
+    'foam.lang.X',
+    'java.util.HashMap',
+    'java.util.List',
+    'java.util.Map',
+    'java.util.Timer'
+  ],
+
+  javaCode: `
+  public PersistentSessionDAO(X x, DAO delegate) {
+    setX(x);
+    setDelegate(delegate);
+  }
+  `,
+
+  properties: [
+    {
+      name: 'persistentDAO',
+      class: 'foam.dao.DAOProperty',
+      javaFactory: 'return new MDAO(Session.getOwnClassInfo());'
+    },
+    {
+      name: 'mDAO',
+      class: 'foam.dao.DAOProperty'
+    },
+    {
+      name: 'journal',
+      class: 'FObjectProperty',
+      of: 'foam.dao.Journal'
+    },
+    {
+      name: 'anonymousCache',
+      class: 'Map',
+      javaFactory: 'return new HashMap();'
+    },
+    {
+      name: 'nullDAO',
+      class: 'foam.dao.DAOProperty',
+      javaFactory: 'return new NullDAO(getX(), Session.getOwnClassInfo());'
+    },
+    {
+      name: 'timerInterval',
+      class: 'Long',
+      value: 1, // 1000,
+      units: 'ms'
+    }
+  ],
+
+  methods: [
+    {
+      documentation: 'Load anonymous users',
+      name: 'start',
+      javaCode: `
+      X x = getX();
+      List<ServiceProvider> providers = (List) ((ArraySink) ((DAO) x.get("serviceProviderDAO")).select(new ArraySink())).getArray();
+      for ( ServiceProvider sp : providers ) {
+        long uid = sp.getAnonymousUser();
+        if ( uid != 0 ) {
+          getAnonymousCache().put(uid, true);
+        }
+      }
+
+      foam.core.logger.Logger logger = foam.core.logger.Loggers.logger(x, this, "start");
+      DAO dao = (DAO) x.get("localSessionDAO");
```

Comment:
To be honest I find this very confusing.  This make PersistentSessionDAO not a very generic decorator, it is now hardwired to expect to find a localSessionDAO service and digs through the whole chain to find an MDAO+JDAO which feels like its breaking encapsuilation.

If the sessionDAO requires such careful construction maybe wer'e better off not using the EasyDAO in the service and just manually configuring the DAO chain so that individual decroators such as this one arn't coupled the cspec configuration and vice versa.